### PR TITLE
Add ETH tick replay research and data pipeline fixes

### PR DESCRIPTION
**Pull request by wuyaocheng**
## Summary

This PR wraps up the ETH research branch and focuses on two things:

1. make `research/backTest.py` support a safer monthly tick replay path that stays aligned with the current enhanced 1-minute backtest logic
2. add the ETH-specific dataset pipeline fixes needed to build and study ETHUSDT data reliably

It also adds a research note documenting the exact test setup, replay logic, and conclusions.

## What changed

- updated `research/backTest.py`
  - replaced the older raw tick scan assumptions with a chunked tick-to-minute event stream replay
  - preserves intraminute `high/low` ordering when replaying monthly tick archives
  - supports both archive formats we encountered:
    - BTC monthly tick files without headers
    - ETH monthly tick files with `id,price,qty,quote_qty,time,...` headers
  - parameterized long/short reentry ATR offsets in the replay paths
- updated `research/ethusdt_1min.py`
  - funding schema compatibility
  - defensive open-interest handling
  - parquet fallback via `pyarrow`
- updated `research/parquet_to_csv_ETHUSDT.py`
  - parquet fallback via `pyarrow`
- added research write-up:
  - `docs/20260413-eth-tick-replay-research.md`

### BTC 2023 control run
Used the new tick replay path as a control to confirm it stays close to the current enhanced 1-minute engine.

Result:
- enhanced_1m: `+69.41%`
- tick_replay: `+67.77%`

Interpretation:
- the new tick replay path is aligned with the current strategy logic
- 1-minute OHLC remains mildly optimistic, but not catastrophically different

### ETH 2023 full-year tick vs 1m comparison
Used the same enhanced configuration for both replay modes:
- `dir2_zero_initial=True`
- `fixed_slippage=0.0005`
- `stop_loss_atr=0.05`
- `stop_mode='atr'`
- `max_trades_per_bar=4`
- `reentry_size_schedule=[0.10, 0.05, 0.025]`
- `trailing_stop_atr=0.3`
- `delayed_trailing_activation=0.5`
- `long_reentry_atr=0.1`
- `short_reentry_atr=0.0`

Result:
- enhanced_1m: `+79.54%`, `273` trades, maxDD `-0.0888%`
- tick_replay: `+65.97%`, `245` trades, maxDD `-0.0830%`

Interpretation:
- the ETH edge still exists under more realistic tick replay
- 1-minute OHLC is still optimistic and overstates annual performance by about `13.58` percentage points in this configuration
- the difference mainly comes from more realistic intraminute path ordering for breakout, whipsaw, and reentry behavior

### Validation performed
- `python3 -m py_compile research/backTest.py research/ethusdt_1min.py research/parquet_to_csv_ETHUSDT.py`
- ETH monthly tick replay smoke on `2023-01`
- ETH full-year `2023 tick vs 1min` run
- BTC 2023 full-year control comparison

## Notes

- The research outputs referenced in the doc are local result artifacts and are not committed in this PR.
- Per repo instructions I also attempted the graph refresh command, but the local environment is currently missing the `graphify` Python module (`ModuleNotFoundError`).

### CODE REVIEW

**Comment by folgercn** [comment]
结论

偏向 Request changes / 至少需要明确修一个点后再过。

不是因为研究方向有问题。
相反，研究思路是对的，文档也写得完整，ETH tick replay vs 1m 这个问题本身很有价值。

我卡的是一个很实际、会直接影响可运行性的问题。

我会拦的点
1. research/backTest.py 里 tick 流索引是 UTC-aware，但信号时间轴很可能是 tz-naive，切片有概率直接炸

这个其实 Codex 也已经指出来了，而且我同意它这个点。

现在 _build_tick_event_stream() 这里：

event_df['timestamp'] = pd.to_datetime(event_df['timestamp_ms'], unit='ms', utc=True)
event_df.set_index('timestamp', inplace=True)

会生成 UTC-aware 的 DatetimeIndex。

而 run_tick_full_scan_dual() 虽然这里加了：

df_4h.index = pd.to_datetime(df_4h.index, utc=True)

单看这个函数本身是自洽的。

但问题是：调用链上其它地方如果还拿 tz-naive 的时间对象来切 df_tick.loc[start_t:end_t]，或者读取后的 df_4h/df_1m 没统一处理，月度 replay 很容易出现：

tz-aware / tz-naive 比较报错
窗口切片为空
或者研究结果 silently 偏掉

而这个 PR 又是研究类 PR，silent misalignment 比直接报错更糟。

为什么我把这个当 blocker

因为这不是“风格问题”，也不是“未来可优化”：

这是可能让 backtest 结果不可信的问题。

研究类代码最怕的就是：

结果看起来跑出来了
但时间轴对齐其实有偏差

这个点不收，我不会放心让它直接进。

我认可这次 PR 的地方
1. tick replay 路径改造方向是对的

从“原始逐 tick 扫描”改成：

chunk 读取
聚成 minute summary
保留 high_ts / low_ts
重建 minute 内事件顺序

这个设计很合理，属于“用更可控的复杂度，换取更真实的 intraminute path”。

2. ETH 数据脚本兼容性补丁是实用的

包括：

funding schema 兼容
OI defensive handling
parquet fallback 到 pyarrow

这些都不是花活，是能明显减少环境摩擦的。

3. 文档写得够完整

docs/20260413-eth-tick-replay-research.md 这份说明把：

配置
试验方法
结果
结论

都写清楚了，这点加分很大。

4. 把 long_reentry_atr / short_reentry_atr 参数化是合理增强

这个比原来统一 REENTRY_ATR 要更适合后续研究。

我看到的其他非阻塞关注点
A. 这个 PR 体量其实有点大

它同时做了：

replay engine 逻辑变化
ETH 数据脚本修复
研究文档补充

放一起能理解，但 review 成本偏高。

B. _build_tick_event_stream() 的事件重建仍然是“minute-level approximation”，不是原始逐 tick replay

这不是问题，只是要确保文档里持续强调：

它比 1m OHLC 更真实
但不是逐 tick 完全还原

从你现在文档看，这点表达还可以。

C. high_ts <= low_ts 的顺序判定在同毫秒场景下仍然有约定性

这个很边缘，不拦。只是如果交易所数据里出现同时间戳多事件，当前 dedupe/order 逻辑仍然是一种近似。

我建议怎么修 blocker

核心就是：把 replay 路径上的时间轴统一到一个明确约定。

最稳的方式有两个方向，二选一：

方案 A：整个 replay 链路统一成 UTC-aware
所有读进来的 df_4h / df_1m / tick event index 都明确 utc=True
所有窗口切片都只用 aware 时间
必要时加 assert

例如进入 replay 前显式检查：

assert df_4h.index.tz is not None
assert df_tick.index.tz is not None
方案 B：统一转成 tz-naive UTC
tick event index 生成后 tz_convert('UTC').tz_localize(None)
signal index 也同样处理成 naive UTC

这样也行，但一定要全链路统一。

我更推荐

方案 A：保留 UTC-aware
因为研究数据本来就是跨来源拼接，保留时区信息更不容易埋坑。

File: research/backTest.py [hunk 1]
```diff
@@ -6,153 +6,534 @@
...
+                continue
+            deduped.append((ts_ms, price, label))
+
+        for seq, (ts_ms, price, label) in enumerate(deduped):
+            events.append((ts_ms, seq, price, label, row.minute_ms))
+
+    event_df = pd.DataFrame(events, columns=['timestamp_ms', 'seq', 'price', 'event', 'minute_ms'])
+    event_df.sort_values(['timestamp_ms', 'seq'], inplace=True)
+    event_df['timestamp'] = pd.to_datetime(event_df['timestamp_ms'], unit='ms', utc=True)
+    event_df.set_index('timestamp', inplace=True)
```
[the remark this answers is not in the document]
Comment (wuyaocheng) [reply research/backTest.py:152]: Addressed in `bc20470`.

This follow-up keeps the replay path on an explicit UTC-aware contract at the engine boundaries instead of relying on whatever `parse_dates=True` produced upstream:
- normalize signal/minute/tick indexes with `pd.to_datetime(..., utc=True)` before slicing
- keep `_build_tick_event_stream()` empty results on a UTC-aware `DatetimeIndex`, so filtered-empty windows do not fall back to a `RangeIndex`

I also smoke-tested the tz-naive signal-index path plus an empty tick-window case locally to make sure the replay no longer trips over naive/aware slicing or empty-window indexing.

**Comment by folgercn** [comment]
Re-checked the latest update.
The earlier timezone-contract concern looks addressed now.

I like the follow-up choice here: instead of relying on whatever `parse_dates=True` happened to produce upstream, the replay boundaries now normalize signal / minute / tick indexes onto an explicit UTC-aware contract, and the empty tick-event case also keeps a UTC-aware `DatetimeIndex`. That removes the biggest reliability concern I had around naive/aware slicing and empty-window behavior.

With that tightened, I do not see a blocker from my side. The tick-to-minute replay refactor looks like a reasonable research-quality compromise, and the ETH data-pipeline fixes plus write-up are useful.

My only remaining note is that this is still a minute-level approximation rather than a full raw-tick replay, so I’d keep that limitation explicit in the docs and in how we interpret the results.
